Standalone secrets page

jhadvig opened the pull request:
@spadgett FYI here is the work I done sofar on the stadalone secrets page + adding select-boxes to add secrets into the BC that will be generated in the `From Image` page. The only two thing I'm missing is the select boxes for adding secrets into the DC on the `From Image` && `Deploy Image` pages, since I need some work from https://github.com/openshift/origin-web-console/pull/656 to be merged, mainly the ability to add multiple secrets with the `osc-secrets` directive. Once thats in, will update the this PR

jhadvig commented on app/views/directives/deploy-image.html:
```diff
@@ -167,8 +176,6 @@
             help-text="Each label is applied to each created resource.">
         </label-editor>
 
-        <alerts alerts="alerts"></alerts>
```
Removing this alert cause we alreadty have `alerts` directive in view https://github.com/openshift/origin-web-console/blob/master/app/views/create.html#L13

jhadvig commented on app/views/directives/from-file.html:
```diff
@@ -26,8 +26,6 @@
         }" ng-model="editorContent" class="editor ace-bordered yaml-mode" id="add-component-editor" required></div>
       </div>
 
-      <alerts alerts="alerts"></alerts>
```
Removing this alert cause we alreadty have `alerts` directive in view https://github.com/openshift/origin-web-console/blob/master/app/views/create.html#L13

spadgett commented on app/views/create/fromimage.html:
```diff
@@ -122,6 +122,18 @@
+                                          secrets-by-type="secretsByType"
+                                          alerts="alerts"
+                                          allow-multiple-secrets="false">
+                              </osc-secrets>
```
This should be below context dir for consistency with the build config editor.

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -9,290 +9,288 @@
     </ui-select>
+    <span ng-if="!newSecret.type" class="help-block">
+      Pick secret type based on it's purpose.
+    </span>
```
I'm not sure how helpful this is. I'd remove the help-block.

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -9,290 +9,288 @@
+        <div class="checkbox">
+          <label>
+            <input type="checkbox" ng-model="newSecret.linkSecret">
+            Link this secret to a Service Account.
```
"Link secret to a service account."

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -9,290 +9,288 @@
+          <label>
+            <input type="checkbox" ng-model="newSecret.linkSecret">
+            Link this secret to a Service Account.
+            <a href="{{'managing_secrets' | helpLink}}" target="_blank"><span class="learn-more-inline">Learn more&nbsp;<i class="fa fa-external-link" aria-hidden="true"></i></span></a>
```
I think this would be better in a learn-more block below the checkbox.

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -9,290 +9,288 @@
-        <div class="help-block">
-          Linking a secret enables a service account to automatically use that secret for some forms of authentication.
-          <a href="{{'managing_secrets' | helpLink}}" target="_blank"><span class="learn-more-inline">Learn more&nbsp;<i class="fa fa-external-link" aria-hidden="true"></i></span></a>
+          <label for="serviceAccount" class="required">Service Account</label>
```
`for` attr but no corresponding `id` attr

spadgett commented on app/views/secrets.html:
```diff
@@ -0,0 +1,81 @@
+                      <th>Created</th>
+                    </tr>
+                  </thead>
+                  <tbody ng-repeat="secret in secretsByType.image | orderObjectsByDate : true">
```
Order by name?

spadgett commented on app/views/secrets.html:
```diff
@@ -0,0 +1,81 @@
+                  <!-- If there are no deployment configs and no replication controllers owned by a deployment config -->
+                  <tr><td colspan="3"><em>No secrets</em></td></tr>
+                </tbody>
+                <tbody ng-repeat="secret in secretsByType.source">
```
Let's sort the secrets like below.

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -9,290 +9,288 @@
+          model="newSecret.data.cacert"
+          drop-zone-id="cacert"
+          dragging="false"
+          help-text="Upload your ca.cert or  file."
```
Upload your ca.cert file.

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -9,290 +9,288 @@
+          }
+        }" ng-model="newSecret.data.privateKey" class="create-secret-editor ace-bordered" id="private-key-editor" required></div>
+        <div class="help-block">
+          Private SSH key file for SCM servers authentication.
```
Private SSH key file for Git authentication.

spadgett commented:
Header is not always showing up for me:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19561254/e62c84a8-96a5-11e6-8d42-d9cbf0fa01fe.png)

spadgett commented:
Maybe "Reveal secret contents" and "Hide secret contents"?

spadgett commented:
I think we need to update this link text to say advanced source options as well.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19561353/61c3c662-96a6-11e6-8c27-f85d53234169.png)

jhadvig commented on app/views/browse/route.html:
```diff
@@ -6,6 +6,7 @@
     <div class="middle-container">
       <div class="middle-header">
         <div class="container-fluid">
+          {{secretName}}
```
yeah.. will nuke it :boom:

spadgett commented on app/scripts/directives/createSecret.js:
```diff
@@ -55,26 +55,33 @@ angular.module("openshiftConsole")
-          $scope.serviceAccountsNames = _.keys($scope.serviceAccounts);
-        });
+        // List SA only if $scope.serviceAccountToLink is not defined so user has to pick one.
+        if (!$scope.serviceAccountToLink && AuthorizationService.canI('serviceaccounts', 'list')) {
```
Also avoid listing if the user can't update service accounts?

spadgett commented on app/scripts/services/secrets.js:
```diff
@@ -14,18 +14,65 @@ angular.module("openshiftConsole")
+        var decodedData = window.atob(encodedData);
+        switch (paramName) {
+          case ".dockercfg":
+              decodedSecretData = decodeDockercfg(decodedData);
```
indentation

spadgett commented on app/scripts/services/secrets.js:
```diff
@@ -14,18 +14,65 @@ angular.module("openshiftConsole")
+              decodedSecretData = decodeDockercfg(decodedData);
+            break;
+          case ".dockerconfigjson":
+              decodedSecretData = decodeDockerconfigjson(decodedData);
```
indentation

spadgett commented on app/views/browse/secret.html:
```diff
@@ -0,0 +1,146 @@
+
+                </dl>
+              </div>
+              <a href="" ng-click="view.showSecret = !view.showSecret">{{view.showSecret ? "Hide" : "Reveal"}} secret content</a>
```
secret contents

spadgett commented on app/views/create-secret.html:
```diff
@@ -0,0 +1,38 @@
+                <div class="mar-top-xl">
+                  <h1>Create Secret</h1>
+                  <div>
+                    <span class="help-block">
```
Nit: should be able to put `class="help-block"` on the div and get rid of the inner span

spadgett commented:
Can we preselect the first secret type in the create form? It looks a little strange with only a select when the form is first displayed.

spadgett commented on app/views/directives/create-secret.html:
```diff
@@ -11,290 +11,287 @@
-              </span>
-            </label>
-          </div>
+      <div class="form-group">
```
Needs a `canI` check for updating service accounts?

jhadvig replied:
The `canI`  check is on line 270

jhadvig commented:
@spadgett comments addressed :) PTAL

spadgett commented on app/views/browse/secret.html:
```diff
@@ -0,0 +1,146 @@
+
+                </dl>
+              </div>
+              <a href="" ng-click="view.showSecret = !view.showSecret">{{view.showSecret ? "Hide" : "Reveal"}} secret contents</a>
```
We should add a div with some bottom margin or padding.

spadgett commented:
Some secrets show no content for me:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/19579339/19457a82-96ee-11e6-8c48-fa9fb81ec69a.png)

jhadvig commented:
@spadgett I think whats the issue in your last comment. When you check the json representation of that secret in CLI you will probably see that the `data` field contains something like:
```
    "data": {
        "ca.cert": "ZXlKaGJHY2lPaUpTV....
```

Correctly it should be `ca.crt`, not `ca.cert`. I had a typo in the creation of the secret, and unfortunately k8s is not doing any validation upon the secrets. There are just few types of secret that they support, rest will take part in the secret data, but actually wont do anything. 
I fixed the issue in my "addressing comments 1" commit, so if you created that secret before that commit you will see the issue.

spadgett commented:
This isn't a secret I created. I think we need to be showing secret data keys, values as `****` until revealed. We should only special case the kinds we know about.

```json
{
  "kind": "Secret",
  "apiVersion": "v1",
  "metadata": {
    "name": "hawkular-cassandra-certificate",
    "namespace": "openshift-infra",
    "selfLink": "/api/v1/namespaces/openshift-infra/secrets/hawkular-cassandra-certificate",
    "uid": "7dcc115c-9460-11e6-b92c-080027242396",
    "resourceVersion": "680",
    "creationTimestamp": "2016-10-17T11:54:42Z",
    "labels": {
      "metrics-infra": "hawkular-cassandra"
    }
  },
  "data": {
    "cassandra-ca.certificate": "***",
    "cassandra.certificate": "***"
  },
  "type": "Opaque"
}
```

spadgett commented on app/views/browse/secret.html:
```diff
@@ -0,0 +1,152 @@
+
+                      <div ng-switch-default>
+                        <dt>{{secretDataName}}:</dt>
+                        <dd ng-if="view.showSecret" class="gutter-bottom">{{secretData}}</dd>
```
`class="word-break"`

spadgett commented on app/scripts/services/secrets.js:
```diff
@@ -53,22 +53,25 @@ angular.module("openshiftConsole")
-      _.each(secretData, function(encodedData, paramName) {
-        var decodedData = window.atob(encodedData);
+      return _.mapValues(secretData, function(data, paramName) {
+        var decodedData = window.atob(data);
```
I'd rather not call `window.atob` until we know that we need to decode the data

spadgett replied:
Maybe

```js
      return _.mapValues(secretData, function(data, paramName) {
        switch (paramName) {
          case ".dockercfg":
            return decodeDockercfg(window.atob(data));
          case ".dockerconfigjson":
            return decodeDockerconfigjson(window.atob(data));
          case "username":
          case "password":
          case ".gitconfig":
          case "ssh-privatekey":
          case "ca.crt":
            return window.atob(data);
          default:
            return data;
        }
      });
```

Might be better to call `window.atob()` inside `decodeDockercfg` and `decodeDockerconfigjson`. Decode is in the function name, which implies to me it will handle that for me.

spadgett commented:
[merge]

spadgett commented:
[merge]

spadgett commented:
Merged as #702
